### PR TITLE
[SPARK-42343][CORE] Ignore `IOException` in `handleBlockRemovalFailure` if SparkContext is stopped

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to suppress verbose `IOException` warnings in `BlockManagerMasterEndpoint.handleBlockRemovalFailure` in case of the stopped `SparkContext`.

### Why are the changes needed?

Although we ignore these kind of exceptions during `SparkContext.stop`, this PR can prevent the misleading error messages like the following.
```
23/02/04 01:26:05 INFO SparkUI: Stopped Spark web UI at http://driver-svc.default.svc:4040
23/02/04 01:26:05 INFO KubernetesClusterSchedulerBackend: Shutting down all executors
23/02/04 01:26:05 INFO KubernetesClusterSchedulerBackend$KubernetesDriverEndpoint: Asking each executor to shut down
23/02/04 01:26:05 WARN ExecutorPodsWatchSnapshotSource: Kubernetes client has been closed.
23/02/04 01:26:05 ERROR TransportResponseHandler: Still have 1 requests outstanding when connection from /172.31.215.139:47148 is closed
23/02/04 01:26:05 WARN BlockManagerMasterEndpoint: Error trying to remove shuffle 0 from block manager BlockManagerId(3, 172.31.215.139, 37477, None)
java.io.IOException: Connection from /172.31.215.139:47148 closed
	at org.apache.spark.network.client.TransportResponseHandler.channelInactive(TransportResponseHandler.java:147)
	at org.apache.spark.network.server.TransportChannelHandler.channelInactive(TransportChannelHandler.java:117)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:81)
	at io.netty.handler.timeout.IdleStateHandler.channelInactive(IdleStateHandler.java:277)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:81)
	at org.apache.spark.network.util.TransportFrameDecoder.channelInactive(TransportFrameDecoder.java:225)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1405)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)
	at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:901)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:831)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

### Does this PR introduce _any_ user-facing change?

Yes, but this PR changes `log` ond and this happens during `SparkContext.stop` and this 

### How was this patch tested?

Manual review.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this please, @mridulm ?

**Comment by mridulm** [comment]
+CC @Ngone51 who last changed this method.

This part of the code has not changed much for couple of years now - any reason why it is showing up as an issue in 3.4 ?

**Comment by dongjoon-hyun** [comment]
Thank you for review, @mridulm . I'm hitting these errors very frequently at `Partitioned 1TB TPCDS` testings for Spark 3.4.
Initially, I suspected several logic changes due to `Push-based Shuffle` PRs, but I realized that the query result is correct and these verbose error messages happen only after stopping `SparkContext`. This is a log-only PR and I want to re-check again after this PR.

**Comment by dongjoon-hyun** [comment]
Also, could you review this, @viirya ?

**Comment by mridulm** [comment]
What I am trying to understand is, is this a regression in 3.4 or was always the case ?

**Comment by viirya** [comment]
Seems not a regression in 3.4. I looked at the code in 3.2, it looks like the same.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 1]
```diff
@@ -231,8 +231,10 @@ class BlockManagerMasterEndpoint(
       bmId: BlockManagerId,
       defaultValue: T): PartialFunction[Throwable, T] = {
     case e: IOException =>
-      logWarning(s"Error trying to remove $blockType $blockId" +
-        s" from block manager $bmId", e)
+      if (!SparkContext.getActive.map(_.isStopped).getOrElse(true)) {
+        logWarning(s"Error trying to remove $blockType $blockId" +
+          s" from block manager $bmId", e)
+      }
       defaultValue
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:238]: Based on the description of the method `handleBlockRemovalFailure`, the change looks okay. The io failure is non-fatal and when SparkContext is stopped, this log looks unnecessary.

**Comment by mridulm** [comment]
Thanks for confirming this is not a regression @viirya !
If that is the case, two follow up queries come:
* Is this sufficient (a point fix at this location) ? Or are there other codepaths which will trigger something similar ?
* If not sufficient, should we modify it in Logging itself - for the `Throwable` variants, to only log message, and not the stacktrace when spark context is stopped ?

**Comment by dongjoon-hyun** [comment]
Yes, this is not a regression. I filed this JIRA as a bug for all live release branches. However, Apache Spark 3.4 hits this at a significant higher rates. @mridulm .

For now, I have only this code path logs.
> Is this sufficient (a point fix at this location) ? Or are there other codepaths which will trigger something similar ?

For this one, I'm still not quite sure if I understand the proposed way clearly, but we can try it separately if there is a ready-to-go patch, @mridulm .
> If not sufficient, should we modify it in Logging itself - for the Throwable variants, to only log message, and not the stacktrace when spark context is stopped ?

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:238]: Thank you!

**Comment by mridulm** [comment]
If fixing this one occurance is enough to reduce the occurance drastically, let us do that in interest of 3.4
We can create a jira to take a deeper look afterwards, and fix it for all.

Perhaps something like this in future ?
Logging.scala:
```
  private def isSparkContextActive = !SparkContext.getActive.map(_.isStopped).getOrElse(true)

  protected def logWarning(msg: => String, throwable: Throwable): Unit = {
    if (log.isWarnEnabled) {
      if (isSparkContextActive) {
        log.warn(msg, throwable)
      } else {
        log.warn(msg)
      }
    }
  }

  protected def logError(msg: => String, throwable: Throwable): Unit = {
    if (log.isErrorEnabled) {
      if (isSparkContextActive) {
        log.error(msg, throwable)
      } else {
        log.error(msg)
      }
    }
}
```

**Comment by dongjoon-hyun** [comment]
Thanks. Looks like a cleaner solution than this ad-hoc one. I can wait for your PR, @mridulm . Please proceed in that way.

BTW, I'm not sure if there are some Spark test logics which depends on logging (without creating SparkContext itself). Anyway, +1 for your direction.

**Comment by dongjoon-hyun** [comment]
Oh, wait. It seems that your PR still log everything except `throwable`?

**Comment by mridulm** [comment]
Yeah, I was proposing to log anyway - but not with the throwable, to reduce noise.
Let us proceed with this PR for the time being @dongjoon-hyun ... I am not sure if there are legitimate reasons why we might want to log exception stack trace during shutdown, and dont want to cause regression in 3.4 this late !

**Review by mridulm — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you. Yep, I'll apply this only at master/3.4.

**Comment by dongjoon-hyun** [comment]
Also, will get all feedbacks (including reverting) during RC period this month.

**Comment by dongjoon-hyun** [comment]
Merged to master/3.4. Thank you, @mridulm and @viirya .

**Pull request closed without merging**